### ISSUE TITLE
Broken type inference with a typeguard in a if statement

### ISSUE THREAD

**Issue by JeanMeche** [microsoft/TypeScript#53216]
# Bug Report

### 🔎 Search Terms

- type guard
- inference
- if condition

### 🕗 Version & Regression Information

This is a regression. 

Inference was OK on `4.7.4` but broken on `4.8.4` and `4.9.5` (tested on the playground).  

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.9.5#code/KYDwDg9gTgLgBDAnmYcDqUCWMCGAjAG2AB4AVAPjgF44BvAKDibgFopgcATCAOwMTgBtANJxMPOAGtgiCADM4pALoAuRSKUBuegF9t9AMYEcAZxNwAYhAh1Gzdl1784c69TgByVxA-ad9Q2MzOAAhHChbZjgHbj4BPHD3DwSoX10ApBR0LFxCYCsIMIiaDGx8ImICuAAfUPDyfU5gI3DUOQBXHgMYTF4xEwKACgg8ACs1HB5EAEo1EdH+y2t9Dq6evphgExhB7xS1UtyiAqLZuAA3CExOSOZMBUGAQkwB613rFOnpuh04AHo-nAABY4MAoHjmHAEEw2ADu2CBYjkgxeQy+ASie3CdiYALgAD0APy6IA)

### 💻 Code

```typescript
export type Writable<T> = {
    -readonly [K in keyof T]: T[K];
};

class Foo {
    readonly foo = 'foo';
}

class Bar {
    readonly bar = 'bar';
}

type WritableFooBar = Writable<Foo | Bar>;

declare function isFoo(obj: any): obj is Foo;

function test(foobar: WritableFooBar): void {
    if (!isFoo(foobar)) {} // happens also with if(isFoo())

    foobar
    // ^? Foo | Writable<Bar>
}

```
### 🙁 Actual behavior

`foobar` is inference as `Foo | Writable<Bar>`

### 🙂 Expected behavior

`foobar` should be inferred as `Writable<Foo|Bar>` or `WritableFooBar`

**Comment by JeanMeche** [comment]
Actually, this has been fixed with 5.0 but could find the PR that fixed it !

**Comment by RyanCavanaugh** [comment]
#52282
